### DISCUSSION TITLE
Handling output in StartProcess RedirectedStandardOutputHandler in Cake.Build

### DISCUSSION THREAD

**Discussion by DmitryLukyanov** [cake-build/cake#4204] [Q&A] [answered]
I want to call a custom tool with cake.build via `StartProcess`. Let it be `findstr`.
All is good, `StartProcess` allow doing it, but now I want to analyze the searching result.
For this target I configured `RedirectedStandardOutputHandler`:

      settings.RedirectStandardOutput = true;
      settings.RedirectedStandardOutputHandler = (str) =>
      {
          if (str != null)
          {
             isFound = true;
          }
          return str;
      };

it works, but here some weirdness that I can't explain yet:
1. With such configuration, `StartProcess` returns `0` where in the rest of cases it returns `1`. 
2. When I call `findStr` in the powershell (to search some pattern in csproj), I see 2 lines in the output:

        %PATH%>findstr %Pattern% App.csproj
        <Reference Include="%Matched%">
          <HintPath>....\%Matched%.dll</HintPath>

but the output (`str` variable) in cake consist of only a single second line. The first one is missed (So there is no `Reference..` line). 

Can anyone bring some light on this?

**Comment by nils-a** [comment]
@DmitryLukyanov you need to share more of your code. Setting `RedirectedStandardOutputHandler` and/or `RedirectedStandardOutput` has no effect on the exit-Code or the output.

As I have written on [SO](https://stackoverflow.com/a/76959434/180156) (I have updated my answer), the `RedirectedStandardOutputHandler` will be called once for each line of output.

So, given the following script:

```csharp
var target = Argument("target", "Default");
var settings = new ProcessSettings
{
 Arguments = "/I task build.cake"
};

Task("NoRedirect")
.Does(() => {
   Information("Running no redirect:");
   var exitCode = StartProcess("findstr", settings);
   Information("exited with code: " + exitCode);
});

Task("Redirect")
.Does(() => {
   Information("Running redirect:");
   settings.RedirectStandardOutput = true;
   settings.RedirectedStandardOutputHandler = (x) => {
      Information("--> " + x);
      return x;
   };
   var exitCode = StartProcess("findstr", settings);
   Information("exited with code: " + exitCode);
});

Task("Default")
   .IsDependentOn("NoRedirect")
   .IsDependentOn("Redirect");

RunTarget(target);
```

Cake produces the following result:
```text
========================================
NoRedirect
========================================
Running no redirect:
Executing: "findstr" /I task build.cake
 Arguments = "/I task build.cake"
Task("NoRedirect")
Task("Redirect")
Task("Default")
exited with code: 0

========================================
Redirect
========================================
Running redirect:
Executing: "findstr" /I task build.cake
-->  Arguments = "/I task build.cake"
--> Task("NoRedirect")
--> Task("Redirect")
--> Task("Default")
-->
exited with code: 0

========================================
Default
========================================

Task                          Duration
--------------------------------------------------
NoRedirect                    00:00:00.0361745
Redirect                      00:00:00.0271926
--------------------------------------------------
Total:                        00:00:00.0646765
```

**Reply by DmitryLukyanov** [reply]
You can see the full code [here](https://github.com/DmitryLukyanov/DependenciesAnalysisPOC/blob/master/Scripts/build.cake#L163) (I see no valuable differences with the already provided). I will check your code and come back

**Reply by DmitryLukyanov** [reply, marked as answer]
it's my mistake. The first output log is "omitted" because it doesn't exist. I used different search patterns: with and without file extension and this is why in one case I saw 2 lines output and in another - a single line.
Regarding the error code, apparently it's related to whether a command can find anything or no. If no - the error code will be 1, otherwise - 0
